tutorial/kalman/tutorial-ukf-linear-example.cpp

```diff
@@ -0,0 +1,229 @@
+ * distribution for additional information about the GNU GPL.
+ *
+ * For using ViSP with software that can not be combined with the GNU
+ * GPL, please contact Inria about acquiring a ViSP Professional
+ * Edition License.
+ *
+ * See https://visp.inria.fr for more information.
+ *
+ * This software was developed at:
+ * Inria Rennes - Bretagne Atlantique
+ * Campus Universitaire de Beaulieu
+ * 35042 Rennes Cedex
+ * France
+ *
+ * If you have questions regarding the use of this file, please contact
+ * Inria at visp@inria.fr
+ *
+ * This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
+ * WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+ *
+*****************************************************************************/
+
+/** \example tutorial-ukf-linear-example.cpp
+ * Example of a simple linear use-case of the Unscented Kalman Filter (UKF). Using a UKF
+ * in this case is not necessary, it is done for learning purpous only.
+ *
+ * The system we are interested in is a system moving on a 2D-plane.
+ *
+ * The state vector of the UKF is:
+ *  \f{eqnarray*}{
+        \textbf{x}[0] &=& x \\
+        \textbf{x}[1] &=& \dot{x} \\
+        \textbf{x}[1] &=& y \\
+        \textbf{x}[2] &=& \dot{y}
+   \f}
+
+ * The measurement \f$ \textbf{z} \f$ corresponds to the position along the x-axis
+ * and y-axis. The measurement vector can be written as:
+ * \f{eqnarray*}{
+        \textbf{z}[0] &=& x \\
+        \textbf{z}[1] &=& y
+   \f}
+
+ * Some noise is added to the measurement vector to simulate a sensor which is
+ * not perfect.
+*/
+
+// UKF includes
+#include <visp3/core/vpUKSigmaDrawerMerwe.h>
+#include <visp3/core/vpUnscentedKalman.h>
+
+// ViSP includes
+#include <visp3/core/vpConfig.h>
+#include <visp3/core/vpGaussRand.h>
+#ifdef VISP_HAVE_DISPLAY
+#include <visp3/gui/vpPlot.h>
+#endif
+
+#if (VISP_CXX_STANDARD >= VISP_CXX_STANDARD_11)
+/**
+ * \brief The process function, that updates the prior.
+ *
+ * \param[in] chi A sigma point.
+ * \param[in] dt The period.
+ * \return vpColVector The sigma point projected in the future.
+ */
+vpColVector fx(const vpColVector &chi, const double &dt)
+{
+  vpColVector point(4);
+  point[0] = chi[1] * dt + chi[0];
+  point[1] = chi[1];
+  point[2] = chi[3] * dt + chi[2];
+  point[3] = chi[3];
+  return point;
+}
+
+/**
+ * \brief The measurement function, that project the prior in the measurement space.
+ *
+ * \param[in] chi The prior.
+ * \return vpColVector The prior projected in the measurement space.
+ */
+vpColVector hx(const vpColVector &chi)
+{
+  vpColVector point(2);
+  point[0] = chi[0];
+  point[1] = chi[2];
+  return point;
+}
+
+int main(/*const int argc, const char *argv[]*/)
+{
+  const double dt = 1.; // Period of 1s
+  const double gt_dx = 1.; // Ground truth displacement along x axis between two measurements
+  const double gt_dy = 0.5; // Ground truth displacement along x axis between two measurements
```

Comment:
Values are not realistic for a robotics system like a mobile robot. Instead of 1 sec, I suggest 0.010 sec, instead of 1. meter, I suggest 0.01 meter displacement along x and 0.005 meter along y

Reply:
Done in 75aedad45a73e122e6238441cfa956e3caa4bd16